### PR TITLE
Send event on unhandled exceptions (alternate)

### LINKED ISSUE #3946: Add event for unexpected error
Our unexpected error handling that brings up a dialog prevents VS from crashing, which means no Watson crash report. We need a way to measure the impact of these errors (not everyone files an issue when they see the dialog).

### PR DESCRIPTION
**Pull request by huguesv**
Fix #3946 Add event for unexpected error

This one goes through our wrappers, unlike https://github.com/Microsoft/PTVS/pull/3963

It also bypasses watson sampling as suggested by telemetry folks.

Added several missing `HandleAllExceptions()` calls in cookiecutter.

### CODE REVIEW

File: Python/Product/Cookiecutter/Telemetry/StringTelemetryRecorder.cs [hunk 1]
```diff
@@ -47,6 +47,10 @@ internal sealed class StringTelemetryRecorder : ITelemetryRecorder, ITelemetryLo
                 }
             }
         }
+
+        public void RecordFault(string eventName, Exception ex, string description, bool dumpProcess) {
```
Comment (AlexanderSher) [inline Python/Product/Cookiecutter/Telemetry/StringTelemetryRecorder.cs]: Why is it empty?

File: Python/Product/Cookiecutter/CookiecutterPackage.cs [hunk 2]
```diff
@@ -89,6 +89,8 @@ public sealed class CookiecutterPackage : Package, IOleCommandTarget {
             _projectSystem = new ProjectSystemClient(DTE);
 
             CookiecutterTelemetry.Initialize();
+
+            VSTaskExtensions._telemetry = CookiecutterTelemetry.Current;
```
Comment (AlexanderSher) [inline Python/Product/Cookiecutter/CookiecutterPackage.cs]: What's the value of `_telemetry`? `CookiecutterTelemetry.Current` is accessible from `VSTaskExtensions`

File: Python/Tests/CookiecutterTests/TestTelemetryRecorder.cs [hunk 3]
```diff
@@ -41,6 +42,11 @@ internal sealed class TestTelemetryRecorder : ITelemetryRecorder, ITelemetryTest
                 }
             }
         }
+
+        public void RecordFault(string eventName, Exception ex, string description, bool dumpProcess) {
+            throw ex;
```
Comment (AlexanderSher) [inline Python/Tests/CookiecutterTests/TestTelemetryRecorder.cs]: `ExceptionDispatchInfo.Capture(ex).Throw()`

File: Python/Product/VSCommon/Infrastructure/VSTaskExtensions.cs [hunk 4]
```diff
@@ -79,20 +82,28 @@ static class VSTaskExtensions {
...
-                        if (site != null) {
-                            TaskDialog.ForException(site, ex, issueTrackerUrl: Strings.IssueTrackerUrl).ShowModal();
-                        } else if (!string.IsNullOrEmpty(logFile)) {
-                            MessageBox.Show(Strings.SeeActivityLog.FormatUI(logFile), Strings.ProductTitle);
-                        }
-                    }
+            bool alreadySeen = true;
+            lock (_displayedMessages) {
+                var key = "{0}:{1}:{2}".FormatInvariant(callerFile, callerLineNumber, ex.GetType().Name);
+                if (_displayedMessages.Add(key)) {
```
Comment (AlexanderSher) [inline Python/Product/VSCommon/Infrastructure/VSTaskExtensions.cs]: only `_displayedMessages.Add(key)` requires lock:
```cs
var key = "{0}:{1}:{2}".FormatInvariant(callerFile, callerLineNumber, ex.GetType().Name);
bool alreadySeen;
lock (_displayedMessages) {
    alreadySeen = !_displayedMessages.Add(key);
}
```

File: Python/Product/VSCommon/Infrastructure/VSTaskExtensions.cs [hunk 5]
```diff
@@ -79,20 +82,28 @@ static class VSTaskExtensions {
...
+                    alreadySeen = false;
+                }
+            }
+
+            Debug.Assert(_logger != null);
```
Comment (AlexanderSher) [inline Python/Product/VSCommon/Infrastructure/VSTaskExtensions.cs]: Do we set this in tests?

File: Python/Product/Cookiecutter/View/CookiecutterSearchPage.xaml.cs [hunk 6]
```diff
@@ -162,15 +162,15 @@ internal partial class CookiecutterSearchPage : Page {
                 return;
             }
 
-            ViewModel.LoadTemplateAsync().DoNotWait();
+            ViewModel.LoadTemplateAsync().HandleAllExceptions(null, GetType()).DoNotWait();
         }
 
         public void UpdateTemplate() {
-            ViewModel.UpdateTemplateAsync().DoNotWait();
+            ViewModel.UpdateTemplateAsync().HandleAllExceptions(null, GetType()).HandleAllExceptions(null, GetType()).DoNotWait();
```
Comment (AlexanderSher) [inline Python/Product/Cookiecutter/View/CookiecutterSearchPage.xaml.cs]: `HandleAllExceptions` is called twice.

File: Python/Product/Cookiecutter/CookiecutterPackage.cs (hunk 2 above)
Comment (huguesv) [reply Python/Product/Cookiecutter/CookiecutterPackage.cs]: That was the reason, but might as well just use one global instead of two.

**Comment by huguesv** [comment]
@zooba is this ok with you?

**Pull request merged**